vendor/bat-native-ads/src/bat/ads/internal/ml/pipeline/pipeline_util.cc

```diff
@@ -0,0 +1,220 @@
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "base/json/json_reader.h"
+#include "bat/ads/internal/ml/data/vector_data.h"
+#include "bat/ads/internal/ml/ml_aliases.h"
+#include "bat/ads/internal/ml/ml_transformation_util.h"
+#include "bat/ads/internal/ml/pipeline/pipeline_info.h"
+#include "bat/ads/internal/ml/pipeline/pipeline_util.h"
+
+namespace ads {
+namespace ml {
+namespace pipeline {
+
+base::Optional<TransformationVector> ParseTransformationsJSON(
+    base::Value* transformations_value) {
+  if (!transformations_value || !transformations_value->is_list()) {
+    return base::nullopt;
+  }
+
+  base::Optional<TransformationVector> transformations = TransformationVector();
+  auto transformation_list = transformations_value->GetList();
+  for (size_t i = 0; i < transformation_list.size(); i++) {
+    const base::Value& transformation = transformation_list[i];
+
+    const std::string* transformation_type =
+        transformation.FindStringKey("transformation_type");
+
+    if (!transformation_type) {
+      return base::nullopt;
+    }
+
+    std::string parsed_transformation_type = *transformation_type;
+
+    if (parsed_transformation_type.compare("TO_LOWER") == 0) {
+      transformations.value().push_back(
+          std::make_unique<LowercaseTransformation>(LowercaseTransformation()));
+    }
+
+    if (parsed_transformation_type.compare("NORMALIZE") == 0) {
+      transformations.value().push_back(
+          std::make_unique<NormalizationTransformation>(
+              NormalizationTransformation()));
+    }
+
+    if (parsed_transformation_type.compare("HASHED_NGRAMS") == 0) {
+      const base::Value* transformation_params =
+          transformation.FindKey("params");
+
+      if (!transformation_params) {
+        return base::nullopt;
+      }
+
+      const base::Optional<int> nb =
+          transformation_params->FindIntKey("num_buckets");
+      if (!nb.has_value()) {
+        return base::nullopt;
+      }
+
+      int num_buckets = nb.value();
+
+      const base::Value* n_gram_sizes =
+          transformation_params->FindListKey("ngrams_range");
+
+      if (!n_gram_sizes) {
+        return base::nullopt;
+      }
+
+      std::vector<int> ngram_range;
+      auto ngram_range_list = n_gram_sizes->GetList();
+      for (size_t i = 0; i < ngram_range_list.size(); i++) {
+        const base::Value& n = ngram_range_list[i];
+        ngram_range.push_back(n.GetInt());
+      }
+      HashedNGramsTransformation hashed_ngrams(num_buckets, ngram_range);
+      transformations.value().push_back(
+          std::make_unique<HashedNGramsTransformation>(hashed_ngrams));
+    }
+  }
+
+  return transformations;
+}
+
+base::Optional<model::Linear> ParseClassifierJSON(
```

Comment:
This is not parsing JSON, can we change JSON to the type of object we are parsing please. Check elsewhere too

Reply:
Done. Renamed to `ParsePipelineTransformations` and `ParsePipelineClassifier`.